Added option to add rotated Box obstacle to TileCache.

aymarfisherman opened the pull request:
Added option to add rotated Box obstacle to TileCache (without messing with the existing box).

Very useful when you have a obstacle with a big difference in x to z ratio and that could be rotated (not aligned), like a wall.

jakobbotsch commented on DetourTileCache/Include/DetourTileCache.h:
```diff
@@ -38,6 +38,7 @@ enum ObstacleState
 enum ObstacleType
 {
 	DT_OBSTACLE_CYLINDER,
+	DT_OBSTACLE_AABB,
```
Please rename this back and use a new name for your addition.
For example: `DT_OBSTACLE_ORIENTED_BOX`.
To make it more clear you can add some comments that `DT_OBSTACLE_BOX` is an AABB and `DT_OBSTACLE_ORIENTED_BOX` is an OBB.

jakobbotsch commented on DetourTileCache/Include/DetourTileCache.h:
```diff
@@ -48,18 +49,26 @@ struct dtObstacleCylinder
 };
 
-struct dtObstacleBox
+struct dtObstacleAabb
```
Undo this renaming please.

jakobbotsch commented on DetourTileCache/Include/DetourTileCache.h:
```diff
@@ -48,18 +49,26 @@ struct dtObstacleCylinder
+	float center[ 3 ];
+	float halfextents[ 3 ];
+	float rotaux[ 2 ]; //{ cos(0.5f*angle)*sin(-0.5f*angle); cos(0.5f*angle)*cos(0.5f*angle) - 0.5 }
+};
```
Please call these `halfExtents` and `rotAux`.

jakobbotsch commented on DetourTileCache/Include/DetourTileCache.h:
```diff
@@ -48,18 +49,26 @@ struct dtObstacleCylinder
 	union
 	{
 		dtObstacleCylinder cylinder;
+		dtObstacleAabb aabb;
```
Rename this back and call the new one something like `orientedBox` or `obox`.

aymarfisherman commented:
I'm glad I could contribute.
I did what you requested.
Let me know if you spot anything else I could do to improve this PR.

jakobbotsch commented on DetourTileCache/Include/DetourTileCache.h:
```diff
@@ -130,8 +139,14 @@ class dtTileCache
 	dtStatus addBoxObstacle(const float* bmin, const float* bmax, dtObstacleRef* result);
+
+	// Obb obstacle: can be rotated in Y.
+	dtStatus addBoxObstacle(const float* bmin, const float* bmax, const float yRadians, dtObstacleRef* result);
```
`bmin`/`bmax` here should be `center` and `halfExtents`. Though can we rename `halfExtents` to just `extents`? This is the convention used in other places (eg `findNearestPoly`). To make it more clear, here is an example: assuming the rotation is 0, passing center = `0, 0, 0` and `extents = 1, 1, 1` should mark the box from `-1, -1, -1` to `1, 1, 1`.

jakobbotsch commented on DetourTileCache/Include/DetourTileCacheBuilder.h:
```diff
@@ -130,6 +130,9 @@ dtStatus dtMarkCylinderArea(dtTileCacheLayer& layer, const float* orig, const fl
 					   const float* bmin, const float* bmax, const unsigned char areaId);
 
+dtStatus dtMarkBoxArea(dtTileCacheLayer& layer, const float* orig, const float cs, const float ch,
+					   const float* center, const float* halfExtents, const float* rotAux, const unsigned char areaId);
```
`extents` please, as used in other places in Detour.

jakobbotsch commented on DetourTileCache/Include/DetourTileCache.h:
```diff
@@ -54,13 +55,21 @@ struct dtObstacleBox
+struct dtObstacleOrientedBox
+{
+	float center[ 3 ];
+	float halfExtents[ 3 ];
```
Just `extents` please (see my comment below).

jakobbotsch commented on DetourTileCache/Source/DetourTileCache.cpp:
```diff
@@ -420,6 +420,45 @@ dtStatus dtTileCache::addBoxObstacle(const float* bmin, const float* bmax, dtObs
 	return DT_SUCCESS;
 }
 
+dtStatus dtTileCache::addBoxObstacle(const float* center, const float* halfExtents, const float yRadians, dtObstacleRef* result)
```
`extents`

jakobbotsch commented on DetourTileCache/Source/DetourTileCacheBuilder.cpp:
```diff
@@ -2041,6 +2041,60 @@ dtStatus dtMarkBoxArea(dtTileCacheLayer& layer, const float* orig, const float c
 }
 
+dtStatus dtMarkBoxArea(dtTileCacheLayer& layer, const float* orig, const float cs, const float ch,
+					   const float* center, const float* halfExtents, const float* rotAux, const unsigned char areaId)
```
`extents`

aymarfisherman commented:
I changed it to 'extents' and fixed the argument naming mismatch between .h and .cpp.

Though I'd like to leave a suggestion: 

Please, consider renaming all 'extents' to 'halfExtents', since 'extents' is misleading and ambiguous, especially when 'halfExtents' is  widely used in the industry. It's expected that people will pass 2x'halfExtents' if you ask for 'extents'.

It happened to me and maybe to other people: I thought 'extents' were the actual the full 'extents' of the box and I was wrongly using the library for years, passing 2.0*halfExtents everywhere it asked for 'extents'. I found out by accident (when implementing this PR) that the 'extents' variable actually meant 'halfExtents'.

If not for those reasons, you should still consider this change for shear disambiguation.

jakobbotsch commented:
Thanks! I am not opposed to renaming `extents` to `halfExtents` everywhere in the library. Would you mind submitting a PR for that?
